extension/wasm/wasm_bindings.cpp

```
@@ -0,0 +1,561 @@
+#define THROW_IF_FALSE(cond, message, ...)           \
+  ({                                                 \
+    if ET_UNLIKELY (!(cond)) {                       \
+      THROW_JS_ERROR(Error, message, ##__VA_ARGS__); \
+    }                                                \
+  })
+
+using namespace emscripten;
+using executorch::aten::ScalarType;
+using executorch::aten::Tensor;
+using ::executorch::extension::BufferDataLoader;
+using ::executorch::runtime::Error;
+using ::executorch::runtime::EValue;
+using ::executorch::runtime::Result;
+using ::executorch::runtime::Tag;
+using ::executorch::runtime::TensorInfo;
+
+namespace executorch {
+namespace extension {
+namespace wasm {
+
+namespace {
+
+// val represents all JS values. Using val_array to specify that we specifically
+// want an array.
+template <typename T>
+using val_array = val;
+
+template <typename T>
+inline void js_array_push(val_array<T>& array, const T& value) {
+  array.call<void>("push", value);
+}
+
+#define JS_FORALL_SUPPORTED_TENSOR_TYPES(_) \
+  _(float, Float)                           \
+  _(int64_t, Long)
+
+inline ssize_t compute_expected_numel(
+    const std::vector<torch::executor::Tensor::SizesType>& sizes) {
+  return executorch::aten::compute_numel(sizes.data(), sizes.size());
+}
+
+template <typename T>
+inline void assert_valid_numel(
+    const std::vector<T>& data,
+    const std::vector<torch::executor::Tensor::SizesType>& sizes) {
+  auto computed_numel = compute_expected_numel(sizes);
+  THROW_IF_FALSE(
+      data.size() >= computed_numel,
+      "Required %ld elements, given %ld",
+      computed_numel,
+      data.size());
+}
+
+template <typename T>
+std::vector<T> convertJSGeneratorToNumberVector(val generator) {
+  std::vector<T> data;
+  while (true) {
+    val next = generator.call<val>("next");
+    if (next["done"].as<bool>()) {
+      break;
+    }
+    data.push_back(next["value"].as<T>());
+  }
```

if data becomes too large, maybe throw an error? 

```
  if (data.size() >= MAX_ELEMENTS) {
    THROW_JS_ERROR("Generator exceeded maximum element count of %zu", MAX_ELEMENTS);
  }
  ```